### ISSUE TITLE
Support for preselect property of CompletionItem

### ISSUE THREAD

**Issue by oblitum** [rust-analyzer/rust-analyzer#2037]
As reference, gopls is one language server that supports it, it's quite useful.

**Comment by kjeremy** [comment]
How does this work in gopls? Is the most likely candidate marked as preselected?

**Comment by oblitum** [comment]
@kjeremy yes, which avoids having to manually select the first menu option most of the time. Just need to confirm on half word.

**Comment by matklad** [comment]
Yeah, we don't support any kind of completion scoring/ordering at the moment, that's a big topic!

**Comment by oblitum** [comment]
OK. Just a reminder that this case refer to just a single most relevant word, not score/ordering of all.

**Comment by kjeremy** [comment]
@oblitum what does that mean exactly? Can you give some examples?

**Comment by kjeremy** [comment]
@oblitum yes but:

> this case refer to just a single most relevant word

That could mean many different things so I would like to understand what you mean by the above with examples of things that both match and do not.

**Comment by oblitum** [comment]
I don't define what's most relevant. I think preselect, as provided by LSP for server to take advantage of, is of their consideration what to pick as the most relevant item in a result set.

When I use gopls with coc.nvim, coc does its local candidate ranking, but leave any of the items preselected according to gopls, it may even not be the first one in the completion list, it doesn't harm completion, as I as a user may or may not click `<enter>` if I feel the preselected item is relevant, which gopls it is most of the time.

**Comment by kjeremy** [comment]
Okay, I misunderstood. I thought that you had a heuristic for deciding what was most relevant.

**Comment by oblitum** [comment]
The only thing I have to say on that subject is what I pointed before, as it's just a single item, the heuristic, whatever it is, may not need to be implemented by having all the items ranked in order just to return a single one.

**Comment by martskins** [comment]
Would string similarity suffice as a way of identifying the most relevant item? Also, maybe `preselect`ing an item if it's the *only item* in the list is a good middle-ground solution?

**Comment by oblitum** [comment]
For other discussion why `preselect` exists you may check https://github.com/MaskRay/ccls/issues/513#issuecomment-564274619.

**Comment by JoshMcguigan** [comment]
We now have a basic version of completion item sorting (using `sortText`), and set `preselect` when the score is above some threshold. This does result in more than one item with `preselect` set, so I'm not sure how the LSP clients handle this. 

@oblitum is this working how you would expect now?

**Comment by kjeremy** [comment]
The spec says:
```
/**

	 * Select this item when showing.
	 *
	 * *Note* that only one completion item can be selected and that the
	 * tool / client decides which item that is. The rule is that the *first*
	 * item of those that match best is selected.
	 */
	preselect?: boolean;

```

**Comment by oblitum** [comment]
> and set `preselect` _when_ the score is above some threshold
> @oblitum is this working how you would expect now?

@JoshMcguigan It's not working as I expect, but it seems behavior changed:

![gif-2021-03-16-161620](https://user-images.githubusercontent.com/1269815/111367048-0ea53080-8673-11eb-9365-b5986e9a41a0.gif)

In my view the gopls server seems to work ideally, it always have a `preselect`ed option while typing (in green background). With rust-analyzer, only `call` showed up.

**Comment by oblitum** [comment]
Random "call" preselect was fixed in https://github.com/neoclide/coc.nvim/pull/3338. Thanks for the help folks.
